core/server/worker/src/main/java/alluxio/worker/netty/DataServerReadHandler.java

```diff
@@ -418,24 +460,83 @@ private void runInternal() {
           packet = getDataBuffer(mChannel, start, packetSize);
         } catch (Exception e) {
           LOG.error("Failed to read data.", e);
-          replyError(mChannel, Protocol.Status.Code.INTERNAL, "", e);
-          break;
+          setError(mChannel, new Error(e, true, Protocol.Status.Code.INTERNAL));
+          continue;
         }
-        if (packet == null) {
+        if (packet != null) {
+          mLock.lock();
+          try {
+            mPosToQueue += packet.getLength();
```

Comment:
I think it is our pattern to use locks. We always guard it with try-catch so that the lock can be released even when there is a runtime exception.